src/smt/theory_str_noodler/theory_str_noodler.cpp

```diff
@@ -1951,6 +1960,129 @@ namespace smt::noodler {
         this->m_membership_todo.push_back(std::make_tuple(expr_ref(re_constr, m), r, is_true));
     }
 
+    /**
+     * @brief Handle is_digit
+     * 
+     * Translates into equivalence:
+     * is_digit(s) <-> s \in [0-9]
+     * 
+     * @param e str.is_digit(s)
+     * 
+     * TODO: This probably makes is_digit always relevant.
+     */
+    void theory_str_noodler::handle_is_digit(expr *e) {
+        if(axiomatized_persist_terms.contains(e))
+            return;
+        axiomatized_persist_terms.insert(e);
+
+        expr *s = nullptr;
+        VERIFY(m_util_s.str.is_is_digit(e, s));
+        // s \in [0-9]
+        expr *s_in_digit = m_util_s.re.mk_in_re(s, m_util_s.re.mk_range(m_util_s.str.mk_string("0"), m_util_s.str.mk_string("9")));
+        // is_digit(s) -> s \in [0-9]
+        add_axiom({~mk_literal(e), mk_literal(s_in_digit)});
+        // ~is_digit(s) -> ~(s \in [0-9])
+        add_axiom({mk_literal(e), ~mk_literal(s_in_digit)});
+    }
+
+    /**
+     * @brief Handle to_code, from_code, to_int, from_int
+     * 
+     * Collects (and possibly creates) variables for the argument and result
+     * of the term and puts them in m_tranformation_todo.
+     */
+    void theory_str_noodler::handle_transform(expr *e) {
+        if(axiomatized_persist_terms.contains(e))
+            return;
+        axiomatized_persist_terms.insert(e);
+
+        expr *s = nullptr;
+
+        TransformationType type;
+        std::string name_of_type;
+        if (m_util_s.str.is_to_code(e, s)) {
+            type = TransformationType::TO_CODE;
+            name_of_type = "to_code";
+        } else if (m_util_s.str.is_from_code(e, s)) {
+            type = TransformationType::FROM_CODE;
+            name_of_type = "from_code";
+        } else if (m_util_s.str.is_stoi(e, s)) {
+            type = TransformationType::TO_INT;
+            name_of_type = "to_int";
+        } else if (m_util_s.str.is_itos(e, s)) {
+            type = TransformationType::FROM_INT;
+            name_of_type = "from_int";
+        } else {
+            UNREACHABLE();
+            return;
+        }
+        bool tranforming_from = (type == TransformationType::FROM_CODE || type == TransformationType::FROM_INT);
+
+        // get the var for the argument
+        expr_ref var_for_s(m);
+        if (tranforming_from) {
+            // for from_code and from_int, the argument has integer type, we create a new var for it
+            var_for_s = mk_int_var_fresh(name_of_type + "_argument");
+            add_axiom({mk_literal(m.mk_eq(s, var_for_s))});
+        } else {
+            // for to_code and to_int, the argument has string tyoe, we have to find the variable for it
+            if (m_util_s.str.is_string(s)) {
+                // it seems that Z3 rewriter handles the case where we tranform from string literal, so this should be unreachable
+                UNREACHABLE();
+                return;
```

Comment:
Do we want to keep this in the hands of Z3 rewriter? Would it benefit us to handle it ourselves here?